File: code/frontend/src/services/product-service.ts
import api from "@/lib/axios";
import { CategoryResponse } from "@/types/backend";

export interface ProductImage {
  id: number;
  imageUrl: string;
  isPrimary: boolean;
}

export interface ProductSummary {
  id: number;
  name: string;
  pricePerDay: number;
  status: "PENDING" | "AVAILABLE" | "RENTED" | "UNAVAILABLE" | "BLOCKED";
  category: {
    id: number;
    name: string;
    slug: string;
  };
  primaryImage: string | null;
  createdAt: string;
}

export interface ProductDetail extends ProductSummary {
  description: string;
  depositAmount: number;
  address: string;
  latitude: number | null;
  longitude: number | null;
  owner: {
    id: number;
    fullName: string;
    email: string;
  };
  images: ProductImage[];
  updatedAt: string;
}

export interface CreateProductInput {
  categoryId: number;
  name: string;
  description: string;
  pricePerDay: number;
  depositAmount: number;
  address: string;
  latitude?: number | null;
  longitude?: number | null;
  images: { imageUrl: string; isPrimary: boolean }[];
}

export const productService = {
  getMyProducts: async (page = 0, size = 10): Promise<{ content: ProductSummary[]; totalElements: number }> => {
    const res = await api.get(`/products/my`, {
      params: { page, size },
    });
    return res.data;
  },

  getProductDetail: async (id: number): Promise<ProductDetail> => {
    const res = await api.get(`/products/${id}`);
    return res.data;
  },

  createProduct: async (data: CreateProductInput): Promise<ProductDetail> => {
    const res = await api.post(`/products`, data);
    return res.data;
  },

  updateProduct: async (id: number, data: Partial<CreateProductInput> & { status?: string }): Promise<ProductDetail> => {
    const res = await api.put(`/products/${id}`, data);
    return res.data;
  },

  deleteProduct: async (id: number): Promise<void> => {
    await api.delete(`/products/${id}`);
  },

  // Public listing — all AVAILABLE products (not owner-restricted)
  getAvailableProducts: async (
    page = 0,
    size = 12,
    categoryId?: number
  ): Promise<{ content: ProductSummary[]; totalElements: number; totalPages: number }> => {
    const params: Record<string, unknown> = { page, size };
    if (categoryId) params.categoryId = categoryId;
    const res = await api.get(`/products`, { params });
    return res.data;
  },

  // Public detail — any authenticated user can view
  getPublicProductDetail: async (id: number): Promise<ProductDetail> => {
    const res = await api.get(`/products/public/${id}`);
    return res.data;
  },

  // Public search and filtering
  getPublicProducts: async (params: {
    page?: number;
    size?: number;
    keyword?: string;
    categoryIds?: number[];
    minPrice?: number;
    maxPrice?: number;
    address?: string;
    latitude?: number;
    longitude?: number;
    radius?: number;
    sort?: "newest" | "price_asc" | "price_desc" | "relevant";
  }): Promise<{ content: PublicProductSummaryResponse[]; totalElements: number; totalPages: number }> => {
    const res = await api.get("/products/public", {
      params,
      paramsSerializer: {
        indexes: null,
      },
    });
    return res.data;
  },

  getCategories: async (): Promise<CategoryResponse[]> => {
    const res = await api.get("/categories");
    return res.data;
  },
};

export interface PublicProductSummaryResponse {
  id: number;
  name: string;
  pricePerDay: number;
  depositAmount: number;
  address: string | null;
  categoryName: string;
  ownerFullName: string;
  primaryImageUrl: string | null;
  status: "PENDING" | "AVAILABLE" | "RENTED" | "UNAVAILABLE" | "BLOCKED";
  createdAt: string;
}

File: code/backend/src/main/java/com/ioc/internship/service/ProductService.java
package com.ioc.internship.service;
import com.ioc.internship.dto.request.CreateProductRequest;
import com.ioc.internship.dto.request.UpdateProductRequest;
import com.ioc.internship.dto.response.ProductDetailResponse;
import com.ioc.internship.dto.response.ProductSummaryResponse;
import com.ioc.internship.entity.ProductStatus;
import org.springframework.data.domain.Page;
public interface ProductService {
    ProductDetailResponse createProduct(CreateProductRequest request, String email);
    ProductDetailResponse updateProduct(Long id, UpdateProductRequest request, String email);
    void deleteProduct(Long id, String email);
    Page<ProductSummaryResponse> getMyProducts(String email, ProductStatus status, int page, int size);
    ProductDetailResponse getProductDetail(Long id, String email);
    Page<ProductSummaryResponse> getAllProductsAdmin(ProductStatus status, int page, int size);
    void updateProductStatusAdmin(Long id, com.ioc.internship.dto.request.UpdateProductStatusRequest request);
    void updateMyProductStatus(Long id, com.ioc.internship.dto.request.UpdateMyProductStatusRequest request, String email);
    Page<ProductSummaryResponse> getAvailableProducts(int page, int size, Long categoryId);
    com.ioc.internship.dto.response.PublicProductDetailResponse getPublicProductDetail(Long id);
    java.util.List<com.ioc.internship.dto.response.BlockedDateRangeResponse> getBlockedDates(Long id);
    Page<com.ioc.internship.dto.response.PublicProductSummaryResponse> getPublicProducts(
            int page, int size, String keyword, java.util.List<Long> categoryIds,
            java.math.BigDecimal minPrice, java.math.BigDecimal maxPrice,
            String address, java.math.BigDecimal latitude, java.math.BigDecimal longitude,
            Double radius, String sort);
}